### DISCUSSION TITLE
Silk.NET/docs/opengl/c1/3-hello-texture

### DISCUSSION THREAD

**Discussion by giscus** [dotnet/Silk.NET#2361] [Webpage Comments]
# Silk.NET/docs/opengl/c1/3-hello-texture

Your one-stop-shop for high performance .NET graphics & compute. A .NET Foundation project.

https://dotnet.github.io/Silk.NET/docs/opengl/c1/3-hello-texture.html

**Comment by Hanprogramer** [comment]
Should add more informations for high level utilities like Input, Window Management, Context Management etc.

**Reply by cbordeman** [reply]
So...why iare the X, Y, and Z coords each 0..1 right now?  Why does the object stretch when the window changes size?  How do I view more than X -1 to 1, etc?  

Are these more OpenGL tutorials coming?

**Reply by RustyGrassG** [reply]
From what I understand, they do -1 to 1 so that the objects stay in place when the window resolution changes. It 'normalizes' your window resolution in a way, and that keeps your objects in the same part of the window?? And it makes it so your math doesn't break when you do other stuff. Imagine you have math for a 1080p monitor, that math would change if the resolution was bigger or smaller. This fixes that, I think?

Also, I think you can access the resolution directly using the OpenGL ViewMatrix(don't ask me how that works, I just know it exists)

And there should be a way to change an OpenGL setting directly that makes your objects not stretch/shrink when your window changes sizes, cause in my project they stay where they are at even when I change the window size

And as far as 'viewing beyond -1 to 1', I don't think that's possible because -1 is the bottom OR left side of your window, and 1 is the opposite of that... 

But if you are asking cause you want to make some sort of 2d  platformer or something like that, 'it's not your character that moves around the world, it's the world that moves around your character' is the best way I can describe it...
You basically make a vector 2 'scroll' or 'offset' variable that increases/decreases based on whatever movement system you have, then add that to whatever 'position' variables you have... this is the way I've always done it, idk if there is another way.

Hope this was helpful!

**Reply by cbordeman** [reply]
I'm making an RTS, so units' positions are not relative to any single unit.  I think changing the viewport might be the answer, thanks, will try it!

The single precision "float" type only has a certain amount of precision and range, and you lose most of it when you limit the world to -1.0...+1.0.

**Comment by Flurop** [comment]
Its very hard to under stand were some of the code needs to go cuz its just not told!
Also i finished `Handling texture data in shader` but my image is upsidedown is that normal?

**Reply by Flurop** [reply]
And i get thees errors:
```
warning CS9193: Argument 3 should be a variable because it is passed to a 'ref readonly' parameter
warning CS9193: Argument 3 should be a variable because it is passed to a 'ref readonly' parameter
warning CS9193: Argument 3 should be a variable because it is passed to a 'ref readonly' parameter
warning CS9193: Argument 3 should be a variable because it is passed to a 'ref readonly' parameter
```

**Reply by angry-dango** [reply]
For the upside down problem, you can fix it by changing how positions map to UVs in the vertices array.
More info: https://stackoverflow.com/questions/5316529/opengl-texture-inverted

**Reply by angry-dango** [reply]
For the warnings, use TexParameter instead of TexParameterI. The article says you have to use TexParameterI because TexParameter doesn't accept enums, but you can still cast the enums to int with TexParameter.

**Reply by appina** [reply]
This is expected behaviour related to how StbImage loads images. The original C library exposes a function called `stbi_set_flip_vertically_on_load(int)` which can be used also in StbImageSharp.
Calling `StbImage.stbi_set_flip_vertically_on_load(1);` once before loading any images will make StbImage flip the images on load.

**Comment by RustyGrassG** [comment]
I enjoyed this tutorial series a lot. This is the only documentation i have gone through (including LearnOpenGL) that has actually made sense to me! 

I see that there are several more sections on the GitHub page. Is there any chance more tutorials will be uploaded here in the future??

Thank you again for taking the time to write these FREE tutorials. Much respect.

**Reply by Perksey** [reply]
Hey there! Thanks so much, glad you've had a good experience with them! :)

Both the examples in the repository and the tutorials on the website are community-driven resources written by a small number of external contributors. We always welcome PRs to improve these resources!

**Comment by simtrip** [comment]
I'm confused by the memory layout diagram showing position and stride for the interleaved Position + UV array. Shouldn't the "stride" labels for both these attributes read 20 bytes, not 32?
